lib/web/apiserver.go

```diff
@@ -523,10 +523,20 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 
 		// serve Web UI:
 		if strings.HasPrefix(r.URL.Path, "/web/app") {
+			// we will set our (weak) etag based on the teleport version
+			etag := fmt.Sprintf("W/%s", teleport.Version)
```

Comment:
I have a dumb question, for dev, how do we ensure the browser doesn't cache it?

Reply:
_normally_, local dev is done with `yarn dev` which starts a development server for the frontend so that isn't changed here. If you aren't using the local dev server and your changes are cached, you'll have to (for now) force reload your browser. In chrome-like browsers, you can do this by pressing cmd(or ctrl)+shift+r

Reply:
The easiest way without creating more build files and what not is to just create a timestamp when the process starts and use that for the etag. This ensures that the files are up to date anytime the proxy comes online. For devs, this is no different than having a value at build time and for production this means that generally, it'll restart when someone upgrades. I'm not sure how often processes would be restart in a production environment but it can't be so often that this would mean a couple more cache updates per update cycle.

Reply:
> However, the cache will kick in even in a situation where I'm not a fullstack dev

yes. i may not be even doing any web-related changes at all, but I would expect to see the latest web UI when I am testing. `16.0.0-dev` could be stuck for a few months and I may not notice that I have web UI cached.

Since I reviewed this change, I can `cmd+shift+r` but others may not be aware. 

I won't block this change for this but we should think of something if we can. I haven't approved the PR yet as I want to double check if we can add some UT, if possible.

Reply:
> The easiest way without creating more build files and what not is to just create a timestamp when the process starts and use that for the etag.

We can have multiple Proxy servers

Reply:
True, which would cause someone's cache to update anytime they are load balanced to another proxy. Ok, that idea is out.

Reply:
https://stackoverflow.com/questions/70590805/how-to-embed-a-string-in-the-compiled-binary-at-build-time

I'd ask some Go heads what's the best way to go about this.

Reply:
This would technically address the development issue, but if you have one proxy running on Linux and another on macOS (unlikely but possible), or one proxy running on arm64 and another on amd64 (still unlikely I guess?), then they'd both have different ETags, even if a random string embedded at compile time was added.

Another way to solve this would be to source the value from the environment in CI. Since we're on CI and we know the value won't be repeated, then the value added to ETag could be just the repeated version number really. In dev though we could embed a random string or whatever.

Connect does something like this already where in dev the version of the app is set to `1.0.0-dev`, but in CI we overwrite the version with the actual Teleport version.

https://github.com/gravitational/teleport/blob/fd562794cb1e5b68701435616d58632e811d6c23/Makefile#L615

Reply:
Could `yarn build-ui-e` and `yarn-build-oss` add a file to the relevant output directory with a hash of `app.js`? And then we can read that file for the etag

Would work for development & releases

Reply:
I just checked and atm, Linux and macOS versions of the frontend app have the same hashes. I checked this on 15.1.8. I suppose this should remain true as we don't use any native packages to build the frontend code, so the hashesh _should_ be the same no matter the platform.

Reply:
Ok so, it seems a little hacky.

I was able to create an "apphash" file during our vite build chain but had to hash manually. Apparently, the hash isn't available to us for two reasons. 1, we removed it from the filename so its never generated. 2, if we left it in the file name (and tried to remove later), we are only given a placeholder anyway. Looking at how vite/rollup do the hash, they just pass unaltered chunk names to `hash.update` so thats what I did here. (well, they do that and then slice out like 8 characters or whatever but i just left the whole hash anyway)

In the app server, we try to open this hash file and get its contents. if it isn't there then we will just use the version as a fallback (which is probably good enough anyway as this is mostly for dev work anyway).

If you find a better way to do this, let me know but I think this works well enough (the golang side of it im sure can be done better so open for tips).
https://github.com/gravitational/teleport/pull/39251/commits/e43a05f92ccf004bfddc529f884cff530e70196a

This means our etag will change for any version or any "build" changes. Also, I changed it back to a strong etag as because we are hashing the output code of app.js, that means it should be serving byte for byte the same contents if tags match.

Reply:
if we are calculating some hashes, could we just implement the proper hash calculation for the `etag` instead of using a weak version?

Reply:
We aren't using a weak tag anymore. We are using a version+hash but i guess we don't even need the version anymore. Either way, it's a strong tag now. I think the version is a nice to have included in the string tho, so we can see at an easier glance what version is being served (in case of debugging i suppose)

Reply:
Vite plugin looks fine but let's move it to it's own file

Reply:
@greedy52 no it is not browsable/curable as far as i can tell. we serve only index.html and things inside the `/app` directory. This lives next to index.html and the server doesn't serve anything else in that dir.

i suppose we _could_ make it curlable but i dont really see the need